compiler/rustc_const_eval/src/interpret/discriminant.rs

```diff
@@ -2,12 +2,28 @@
 
 use rustc_middle::mir;
 use rustc_middle::ty::layout::{LayoutOf, PrimitiveExt};
-use rustc_middle::ty::{self, Ty};
+use rustc_middle::ty::{self, ScalarInt, Ty};
 use rustc_target::abi::{self, TagEncoding};
 use rustc_target::abi::{VariantIdx, Variants};
 
 use super::{ImmTy, InterpCx, InterpResult, Machine, Readable, Scalar, Writeable};
 
+/// The tag of an enum discriminant.
+pub(crate) enum Tag {
+    /// No tag; the variant is `Single`-encoded.
+    None,
+    /// The variant is tagged.
+    Tagged { tag: ScalarInt, tag_field: usize },
+    /// No tag; the variant is identified by its validity.
+    Untagged,
```

Comment:
I think `Untagged` is the name you should use for `Single`-encoded variants.

Not sure what to do with the case of "no tagging needed as the value is outside the niche and that's enough". One option would be to just merge that case with `Untagged`. In both cases, there's nothing to be done for writing the tag.

Does the transmutability analysis need to distinguish these?

Reply:
> Does the transmutability analysis need to distinguish these?

Transmutability analysis doesn't, but `write_discriminant` does: it has an extra validity check in the `Untagged` case.